packages/firestore/test/unit/local/simple_db.test.ts

```diff
@@ -481,23 +486,113 @@ describe('SimpleDb', () => {
       ['foo', 'd'],
       ['foob']
     ];
-    await db.runTransaction('readwrite', ['users', 'docs'], txn => {
-      const docsStore = txn.store<string[], string>('docs');
-      return PersistencePromise.waitFor(
-        keys.map(key => {
-          const value = 'doc ' + key.join('/');
-          return docsStore.put(key, value);
-        })
-      );
-    });
+    await db.runTransaction(
+      'readwrite',
+      /* idempotent= */ false,
+      ['users', 'docs'],
+      txn => {
+        const docsStore = txn.store<string[], string>('docs');
+        return PersistencePromise.waitFor(
+          keys.map(key => {
+            const value = 'doc ' + key.join('/');
+            return docsStore.put(key, value);
+          })
+        );
+      }
+    );
 
-    await db.runTransaction('readonly', ['docs'], txn => {
-      const store = txn.store<string[], string>('docs');
-      const range = IDBKeyRange.bound(['foo'], ['foo', 'c']);
-      return store.loadAll(range).next(results => {
-        expect(results).to.deep.equal(['doc foo', 'doc foo/bar/baz']);
-      });
-    });
+    await db.runTransaction(
+      'readonly',
+      /* idempotent= */ false,
+      ['docs'],
+      txn => {
+        const store = txn.store<string[], string>('docs');
+        const range = IDBKeyRange.bound(['foo'], ['foo', 'c']);
+        return store.loadAll(range).next(results => {
+          expect(results).to.deep.equal(['doc foo', 'doc foo/bar/baz']);
+        });
+      }
+    );
+  });
+
+  it('retries idempotent transactions', async () => {
+    let attemptCount = 0;
+
+    const result = await db.runTransaction(
```

Comment:
This function isn't actually idempotent and I can't see from reading this why this would actually fail and trigger multiple attempts. Could you add some comments describing your intent here?

Reply:
I made the failure more obvious by adding a comment pointing out exactly where it would fail.

As for the idempotency: I think it will be mighty hard to write a unit test for an actually idempotent function that fails on the first attempt but succeeds thereafter. I hope that the intent of the code speaks for itself.

Reply:
Got it. I renamed the test to 'retries transactions marked as idempotent' (since 'readwrite-idempotent' is hopefully short lived).